File: 04-05-2024/Construct Binary Tree from Inorder and Postorder.cpp
//{ Driver Code Starts
/* program to construct tree using inorder and postorder traversals */
#include <bits/stdc++.h>
using namespace std;

/* A binary tree node has data, pointer to left child
   and a pointer to right child */
struct Node {
    int data;
    struct Node* left;
    struct Node* right;

    Node(int x) {
        data = x;
        left = right = NULL;
    }
};

/* This funtcion is here just to test buildTreeUtil() */
void preOrder(Node* node) {
    if (node == NULL) return;

    /* then print the data of node */
    printf("%d ", node->data);

    /* first recur on left child */
    preOrder(node->left);

    /* now recur on right child */
    preOrder(node->right);
}


// } Driver Code Ends
/* Tree node structure

struct Node
{
    int data;
    struct Node* left;
    struct Node* right;

    Node(int x){
        data = x;
        left = right = NULL;
    }
};*/
//Your Code Start
class Solution {
public:
    // Function to return a tree created from postorder and inorder traversals.
    Node *buildTreeUtil(int in[], int post[], int inStart, int inEnd, int &postIndex) {
        // If start index is greater than end index, then no subtree exists.
        if (inStart > inEnd)
            return nullptr;

        // Create a new node with the value at postIndex (last element of postorder).
        Node *node = new Node(post[postIndex--]);

        // If the node has no children, return it.
        if (inStart == inEnd)
            return node;

        // Find the index of this node in the inorder traversal.
        int inIndex;
        for (int i = inStart; i <= inEnd; ++i) {
            if (in[i] == node->data) {
                inIndex = i;
                break;
            }
        }

        // Recursively build right subtree first because postorder gives us root node last.
        node->right = buildTreeUtil(in, post, inIndex + 1, inEnd, postIndex);

        // Then recursively build left subtree.
        node->left = buildTreeUtil(in, post, inStart, inIndex - 1, postIndex);

        return node;
    }

    Node *buildTree(int in[], int post[], int n) {
        int postIndex = n - 1; // Start from the last index of postorder array
        return buildTreeUtil(in, post, 0, n - 1, postIndex);
    }

};
//Your Code End



//{ Driver Code Starts.

int main() {
    int t, n;
    cin >> t;
    while (t--) {
        cin >> n;
        int in[n], post[n];
        for (int i = 0; i < n; i++) cin >> in[i];
        for (int i = 0; i < n; i++) cin >> post[i];
        Solution obj;
        Node* root = obj.buildTree(in, post, n);
        preOrder(root);
        cout << endl;
    }
    return 0;
}

// } Driver Code Ends

File: 04-05-2024/Construct Binary Tree from Inorder and Postorder.py
#User function Template for python3

'''
class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None
'''

#Function to return a tree created from postorder and inorder traversals.
#Your Code Start
class Solution:
    def buildTree(self, inOrder, postOrder, n):
        # Index to keep track of the current element of the postorder traversal list
        postIndex = [n - 1]

        # Helper function to construct the tree recursively
        def buildTreeUtil(inStart, inEnd):
            # If start index is greater than end index, then no subtree exists.
            if inStart > inEnd:
                return None

            # Create a new node with the value at postIndex (last element of postorder).
            node = Node(postOrder[postIndex[0]])
            postIndex[0] -= 1

            # If the node has no children, return it.
            if inStart == inEnd:
                return node

            # Find the index of this node in the inorder traversal.
            inIndex = inOrder.index(node.data)

            # Recursively build right subtree first because postorder gives us root node last.
            node.right = buildTreeUtil(inIndex + 1, inEnd)

            # Then recursively build left subtree.
            node.left = buildTreeUtil(inStart, inIndex - 1)

            return node

        # Call the helper function with appropriate parameters
        return buildTreeUtil(0, n - 1)
#Your Code End


#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys
from collections import  defaultdict

#Contributed by : PranchalK


_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())



# Helper function that allocates  
# a new node  
class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None

# This funtcion is here just to test  
def preOrder(node):
    if node == None:
        return
    print(node.data, end=" ")
    preOrder(node.left)
    preOrder(node.right)
    
if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        n = int(input())  # number of nodes in tree
        in_order = list(map(int, input().strip().split()))  # parent child info in list
        post_order = list(map(int, input().strip().split()))  # parent child info in list
        ob = Solution()
        preOrder(ob.buildTree(in_order,post_order,n))
        print()


# } Driver Code Ends
